microbench/src/main/java/io/netty/handler/codec/Base64Benchmark.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.handler.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.handler.codec.base64.Base64Dialect;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.internal.ThreadLocalRandom;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.nio.ByteBuffer;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 10)
+@Measurement(iterations = 20)
+public class Base64Benchmark extends AbstractMicrobenchmark {
+    private final ResourceLeakDetector.Level oldLevel = ResourceLeakDetector.getLevel();
+    private ByteBufAllocator heapAllocator;
+    private ByteBufAllocator directAllocator;
+    private ByteBufAllocator pooledDirectAllocator;
+
+    @Param({ "1024", "65536" })
+    public int size;
+
+    @Param({ "DISABLED", "SIMPLE", "ADVANCED", "PARANOID" })
+    public String level;
+
+    private ByteBuf src;
+
+    @Setup
+    public void setup() {
+        ResourceLeakDetector.setLevel(ResourceLeakDetector.Level.valueOf(level));
+
+        byte[] bytes = new byte[size];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        src = Unpooled.wrappedBuffer(bytes);
```

Comment:
we should test when this is a direct buffer too. The JDK will incur an additional copy because it only operates on heap data.